### PR TITLE
Add preference for full screen navigation drawer

### LINKED ISSUE #8936: [Bug] Full screen navigation draw gesture interferes w/scrolling, non-whiteboard drawing, gestures. Make it optional + default off
Since 2.15, I can't (pretend) draw on the screen anymore when reviewing kanji cards because any left-to-right swipe pulls out the side menu.

Update: Oh, turns out there is a "whiteboard" feature for this.

### PR DESCRIPTION
**Pull request by ShridharGoel**
## Purpose / Description
Add preference for full screen navigation drawer.

## Fixes
Fixes #8936 

## Approach
Added preference for selecting whether full screen navigation drawer should be enabled. By default, it will be off.
 
## How Has This Been Tested?
Verified that the functionality is working.

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by Arthur-Milchior — COMMENTED** [review]
I hope we can find better. This lead to a lot of code repetition. If we have to duplicate layout files for each potential option that can be on or off, it will lead to a lot of code duplication.

I really don't know a lot about views, so I don't know the path forward and would probably require a lot of time to find what should actually be done.

I hope there is a way to create our own layout/view, that takes as children other layouts. Then it behaves either as a standard drawable or as a full drawable depending on a parameter.

File: AnkiDroid/src/main/res/layout-television/reviewer.xml [hunk 1]
```diff
@@ -25,9 +25,6 @@
     <!-- First child of DrawerLayout assumed to be main content, others are drawers.
      We use a Coordinator layout as root View for main content as it allows snackbars
      to be swiped off screen. Other behaviors can also be added in the future if necessary -->
-    <com.drakeet.drawer.FullDraggableContainer
```
[the remark this answers is not in the document]
Comment (ShridharGoel) [reply AnkiDroid/src/main/res/layout-television/reviewer.xml:28]: There are two versions for `layout-television` as well: `reviewer.xml` and `reviewer_with_fullscreen_drawer.xml`.

Should I still revert it?

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 2]
```diff
@@ -953,6 +952,12 @@ protected void onCreate(Bundle savedInstanceState) {
     }
 
     protected int getContentViewAttr(int fullscreenMode) {
+        SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(getBaseContext());
+
+        if (preferences.getBoolean("gestureFullScreenNavigationDrawer", false)) {
```
[the remark this answers is not in the document]
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java]: Updated.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 3]
```diff
@@ -108,11 +98,34 @@
...
+import static com.ichi2.anim.ActivityTransitionAnimation.Direction.FADE;
+import static com.ichi2.anim.ActivityTransitionAnimation.Direction.START;
+import static com.ichi2.anki.CardBrowser.Column.ANSWER;
+import static com.ichi2.anki.CardBrowser.Column.CARD;
+import static com.ichi2.anki.CardBrowser.Column.CHANGED;
+import static com.ichi2.anki.CardBrowser.Column.CREATED;
+import static com.ichi2.anki.CardBrowser.Column.DECK;
+import static com.ichi2.anki.CardBrowser.Column.DUE;
+import static com.ichi2.anki.CardBrowser.Column.EASE;
+import static com.ichi2.anki.CardBrowser.Column.EDITED;
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:120]: There is quite a lot of things here. Can you deactivate the feature that make all of those change? They are unrelated and make hard to read what your commit actually do

**Review by Arthur-Milchior — COMMENTED** [review]
If it's a high priority, I guess we can accept it as is. I'll open an issue to check whether it can be factorized after that

File: AnkiDroid/src/main/res/layout-television/reviewer.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/res/layout-television/reviewer.xml:28]: I don't think you have to revert. I guess it makes sens to use the same thing than everywhere.
But really, I've no idea what it means to have a drawer on TV in the first place

**Comment by ShridharGoel** [comment]
Updated to implement this in a better way.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java (hunk 3 above)
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:123]: I'll revert these changes.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 4]
```diff
@@ -460,7 +461,20 @@ protected void onCreate(Bundle savedInstanceState) throws SQLException {
...
+        setContentView(R.layout.navigation_drawer_layout);
+        ClosableDrawerLayout closableDrawerLayout = findViewById(R.id.drawer_layout);
+        CoordinatorLayout coordinatorLayout = (CoordinatorLayout) LayoutInflater.from(this).inflate(R.layout.homescreen, closableDrawerLayout, false);
+        if (preferences.getBoolean(FULL_SCREEN_NAVIGATION_DRAWER, false)) {
+            FullDraggableContainer fullDraggableContainer = new FullDraggableContainer(this);
+            fullDraggableContainer.addView(coordinatorLayout);
+            closableDrawerLayout.addView(fullDraggableContainer, 0);
+        } else {
+            closableDrawerLayout.addView(coordinatorLayout, 0);
+        }
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:476]: It seems those lines are copy-pasted 6 times. Can you factorize it into a method in `NavigationDrawerActivity`If I get it right, the main difference between each case is the layout ID you are using. If this is right, maybe you can put the code into `super.onCreate` and just make it call an abstract method that returns the Layer Resource

**Comment by david-allison-1** [comment]
Test failures are legit

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 5]
```diff
@@ -158,16 +145,70 @@
...
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_DELETE;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_EDIT;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_EXIT;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_FLIP_OR_ANSWER_BETTER_THAN_RECOMMENDED;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_FLIP_OR_ANSWER_EASE1;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_FLIP_OR_ANSWER_EASE2;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_FLIP_OR_ANSWER_EASE3;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_FLIP_OR_ANSWER_EASE4;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_FLIP_OR_ANSWER_RECOMMENDED;
+import static com.ichi2.anki.cardviewer.ViewerCommand.COMMAND_LOOKUP;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:183]: Could you revert these changes unless necessary, it makes it harder to read the diff

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
If I understand this correctly, it's a large change on the `.xml` side because it's reverting https://github.com/ankidroid/Anki-Android/pull/8071 and implementing it in a much nicer manner.

This is great! It would be much nicer if some work could be done to refactor the call in the code. I think we could handle it in: `NavigationDrawerActivity.setContentView` (if possible).

If not, let us know and we'll hopefully be able to come up with something better

**Comment by ShridharGoel** [comment]
@Arthur-Milchior @david-allison-1 Thanks for the suggestions. I have updated the code.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 6]
```diff
@@ -941,7 +941,7 @@ protected void onCreate(Bundle savedInstanceState) {
         mTagsDialogFactory = new TagsDialogFactory(this).attachToActivity(this);
 
         super.onCreate(savedInstanceState);
-        setContentView(getContentViewAttr(mPrefFullscreenReview));
+        super.setContentView(getContentViewAttr(mPrefFullscreenReview));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:944]: nit: you can keep these as `setContentView(getContentViewAttr(mPrefFullscreenReview));`

File: AnkiDroid/src/main/java/com/ichi2/anki/NavigationDrawerActivity.java [hunk 7]
```diff
@@ -73,12 +77,30 @@
...
     /**
      * runnable that will be executed after the drawer has been closed.
      */
     private Runnable mPendingRunnable;
 
+    @Override
+    public void setContentView(int layoutResID) {
+        SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(getBaseContext());
+
+        ClosableDrawerLayout closableDrawerLayout = (ClosableDrawerLayout) LayoutInflater.from(this).inflate(R.layout.navigation_drawer_layout, null, false);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/NavigationDrawerActivity.java:91]: A comment here explaining why would be useful

**Review by david-allison-1 — COMMENTED** [review]
Looks great! only a couple nitpicks, then good to go

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 6 above)
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java]: Updated.

File: AnkiDroid/src/main/java/com/ichi2/anki/NavigationDrawerActivity.java (hunk 7 above)
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/NavigationDrawerActivity.java:92]: Updated.

**Review by david-allison-1 — APPROVED** [review]
LGTM, thanks!

**Review by mikehardy — APPROVED** [review]
Nice! Thanks @ShridharGoel

**Comment by mikehardy** [comment]
This is the root cause of #9052 - on a phone layout (not fragmented with study options already open on side of deck list) if you touch the 3 digits next to any deck in the deck list it *should* open the study options for that deck but instead it crashes. I'm reverting it on the stable branch.

**Comment by ShridharGoel** [comment]
The crash caused by this has been fixed in #9057.

**Pull request merged**